File: libraries/typescript/src/integrations/patter-tool.ts
/**
 * PatterTool — wrap a live Patter instance as a tool callable from external
 * agent frameworks (OpenAI Assistants, Anthropic Claude tool-use, LangChain,
 * Hermes Agent, MCP, generic OpenAI-compatible endpoints).
 *
 * Pattern this enables: a customer already runs an agent in their existing
 * stack (LangChain, OpenAI Assistant, Hermes Agent, …) and wants the agent
 * to *make phone calls* during a conversation. With this tool, the customer
 * registers `make_phone_call` and the agent's tool-call loop can dial out
 * via Patter, get a transcript + cost back, and continue reasoning.
 *
 * ## Design
 *
 * Each `PatterTool` wraps one `Patter` instance (carrier + agent + serve).
 * The tool exposes:
 *
 *   - `openaiSchema()`     — OpenAI / chat-completions tool spec
 *   - `anthropicSchema()`  — Anthropic Claude tool spec
 *   - `hermesSchema()`     — Hermes Agent / Nous registry schema (alias for
 *                            anthropicSchema; same JSON-Schema shape)
 *   - `execute(args)`      — dial outbound, await call end, return summary
 *   - `hermesHandler()`    — `(args, **kw) => Promise<string>` wrapper that
 *                            returns a JSON string and `{"error": "..."}` on
 *                            failure (matches Hermes' tool contract)
 *
 * ## Usage (OpenAI / Anthropic)
 *
 * ```ts
 * import { Patter, Twilio, DeepgramSTT, GroqLLM, ElevenLabsTTS } from 'getpatter';
 * import { PatterTool } from 'getpatter/integrations';
 *
 * const phone = new Patter({
 *   carrier: new Twilio(),
 *   phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
 *   webhookUrl: 'agent.example.com',
 * });
 *
 * const tool = new PatterTool({
 *   phone,
 *   agent: { stt: new DeepgramSTT(), llm: new GroqLLM(), tts: new ElevenLabsTTS() },
 * });
 *
 * await tool.start();   // boots phone.serve() once
 *
 * // Register with your LLM
 * const tools = [tool.openaiSchema()];
 *
 * // When the LLM emits a tool_call:
 * const result = await tool.execute({
 *   to: '+15551234567',
 *   goal: 'Book a dentist appointment for next Tuesday afternoon.',
 * });
 * // → { call_id, status, duration_seconds, cost_usd, transcript, … }
 * ```
 *
 * ## Usage (Hermes Agent)
 *
 * Hermes' contract: handler takes `args: dict` + kwargs, returns a JSON
 * string. The TS SDK is meant to be invoked from Python via your own bridge
 * (HTTP, MCP, subprocess); this `hermesSchema()` + `hermesHandler()` pair
 * matches the Python adapter shipped under `getpatter.integrations` so the
 * two SDKs stay in lockstep.
 *
 * For pure-Python Hermes setups, use `PatterTool` from `getpatter.integrations`
 * directly inside a `tools/patter.py` module:
 *
 * ```python
 * from tools.registry import registry
 * from getpatter.integrations import PatterTool
 *
 * tool = PatterTool(phone=...)
 * tool.register_hermes(registry)
 * ```
 */

import { EventEmitter } from 'node:events';
import type { Patter } from '../client';
import type { MetricsStore, SSEEvent } from '../dashboard/store';
import type { AgentOptions } from '../types';

/** JSON-Schema of the call args. Identical wire shape across openai/anthropic/hermes. */
const PARAMETERS_SCHEMA = {
  type: 'object' as const,
  properties: {
    to: {
      type: 'string',
      description:
        'Destination phone number in E.164 format (e.g. "+15551234567"). Required.',
    },
    goal: {
      type: 'string',
      description:
        "What the agent should accomplish on the call. Becomes the in-call agent's system prompt for this single call.",
    },
    first_message: {
      type: 'string',
      description:
        'Optional first message the agent speaks when the callee answers. Defaults to a generic greeting.',
    },
    max_duration_sec: {
      type: 'integer',
      description:
        'Hard timeout for the call in seconds. Default 180. The call is force-ended at this deadline whether or not it has resolved.',
      minimum: 5,
      maximum: 1800,
    },
  },
  required: ['to'],
} as const;

const DEFAULT_NAME = 'make_phone_call';
const DEFAULT_DESCRIPTION =
  'Place a real outbound phone call. Returns a JSON object with the full transcript, ' +
  'call status, duration in seconds, and cost. Use this when the user asks you to call ' +
  'someone, schedule appointments by phone, or otherwise reach a human via voice.';

/** Constructor options for `PatterTool`. */
export interface PatterToolOptions {
  /**
   * Patter instance to dial through. Must be in local mode (have a `carrier`).
   * The tool boots `phone.serve()` on `start()`; do not call `serve()` yourself.
   */
  phone: Patter;
  /**
   * Default agent config used for outbound calls. Per-call overrides come from
   * `execute({ goal, first_message })`.
   */
  agent?: AgentOptions;
  /** Tool name shown to the LLM. Default `'make_phone_call'`. */
  name?: string;
  /** Tool description for the LLM. Default tuned for English assistants. */
  description?: string;
  /** Default per-call timeout in seconds. Default 180. */
  maxDurationSec?: number;
  /**
   * Optional pass-through for `phone.serve()`'s `recording` flag — record all
   * outbound calls placed via this tool.
   */
  recording?: boolean;
}

/** Args accepted by `PatterTool.execute()` (and the OpenAI/Anthropic/Hermes tool schemas). */
export interface PatterToolExecuteArgs {
  to: string;
  goal?: string;
  first_message?: string;
  max_duration_sec?: number;
}

/** Result envelope returned by `PatterTool.execute()` once the underlying call ends. */
export interface PatterToolResult {
  call_id: string;
  status: string;
  duration_seconds: number;
  cost_usd?: number;
  transcript: Array<{ role: string; text: string; timestamp?: number }>;
  metrics?: Record<string, unknown> | null;
}

interface PendingCall {
  resolve: (r: PatterToolResult) => void;
  reject: (e: Error) => void;
  timer: NodeJS.Timeout;
  startedAt: number;
}

/** Wraps a live `Patter` instance as a tool callable from external agent frameworks. */
export class PatterTool {
  readonly name: string;
  readonly description: string;
  private readonly phone: Patter;
  private readonly agent: AgentOptions | undefined;
  private readonly maxDurationSec: number;
  private readonly recording: boolean;
  private started = false;
  /** Resolver for the next `call_initiated` SSE event. Only set inside the
   *  dial mutex (`dialQueue`), so two parallel `execute()` calls never share
   *  it and never lose a dispatch. */
  private pendingDial: ((callId: string) => void) | null = null;
  /** Mutex that serializes the dial → call_id capture critical section.
   *  Each `execute()` chains a continuation onto this promise so the
   *  `pendingDial` slot is owned by exactly one caller at a time. */
  private dialQueue: Promise<void> = Promise.resolve();
  /** Captured SSE listener so `stop()` can detach it (prevents leaks when
   *  the underlying Patter instance outlives this tool). */
  private sseListener: ((event: SSEEvent) => void) | null = null;
  /** Captured Patter metrics store, for cleanup in `stop()`. */
  private metricsStoreRef: MetricsStore | null = null;
  /** call_id → pending promise machinery. */
  private readonly pending = new Map<string, PendingCall>();
  private readonly bus = new EventEmitter();
  /** How long to wait for the `call_initiated` SSE before failing the dial. */
  private static readonly DIAL_CAPTURE_TIMEOUT_MS = 10_000;

  constructor(opts: PatterToolOptions) {
    if (!opts.phone) {
      throw new Error('PatterTool: `phone` (a Patter instance) is required.');
    }
    this.phone = opts.phone;
    this.agent = opts.agent;
    this.name = opts.name ?? DEFAULT_NAME;
    this.description = opts.description ?? DEFAULT_DESCRIPTION;
    this.maxDurationSec = Math.max(5, Math.min(1800, opts.maxDurationSec ?? 180));
    this.recording = opts.recording ?? false;
  }

  // --- Schema exporters ---------------------------------------------------

  /** OpenAI Chat Completions / Assistants tool spec. */
  openaiSchema(): {
    type: 'function';
    function: { name: string; description: string; parameters: typeof PARAMETERS_SCHEMA };
  } {
    return {
      type: 'function',
      function: {
        name: this.name,
        description: this.description,
        parameters: PARAMETERS_SCHEMA,
      },
    };
  }

  /** Anthropic Messages API tool spec. */
  anthropicSchema(): {
    name: string;
    description: string;
    input_schema: typeof PARAMETERS_SCHEMA;
  } {
    return {
      name: this.name,
      description: this.description,
      input_schema: PARAMETERS_SCHEMA,
    };
  }

  /**
   * Hermes Agent (Nous Research) registry schema. Same JSON-Schema shape as
   * Anthropic's; Hermes consumes it via `registry.register({ schema: ... })`.
   */
  hermesSchema(): {
    name: string;
    description: string;
    parameters: typeof PARAMETERS_SCHEMA;
  } {
    return {
      name: this.name,
      description: this.description,
      parameters: PARAMETERS_SCHEMA,
    };
  }

  // --- Lifecycle ----------------------------------------------------------

  /** Start the underlying Patter server. Idempotent. */
  async start(): Promise<void> {
    if (this.started) return;
    if (!this.agent) {
      throw new Error(
        'PatterTool.start: `agent` config is required. Pass `{ stt, llm, tts }` ' +
          'or an `engine` (e.g. OpenAIRealtime) when constructing PatterTool.',
      );
    }
    const builtAgent = this.phone.agent(this.agent);
    await this.phone.serve({
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      agent: builtAgent as any,
      recording: this.recording,
      onCallEnd: this.onCallEndHandler.bind(this),
    });

    // Subscribe to the metrics store so we can correlate outbound dials
    // (call_initiated) with the call_id Patter assigns at dial time.
    const store = this.phone.metricsStore;
    if (!store) {
      throw new Error(
        'PatterTool.start: phone.metricsStore is null after serve() — is the dashboard disabled?',
      );
    }
    const listener = (event: SSEEvent) => {
      if (event.type === 'call_initiated' && this.pendingDial) {
        const callId = (event.data.call_id as string) || '';
        if (callId) {
          const dispatch = this.pendingDial;
          this.pendingDial = null;
          dispatch(callId);
        }
      }
    };
    store.on('sse', listener);
    this.sseListener = listener;
    this.metricsStoreRef = store;

    this.started = true;
  }

  /** Stop the underlying Patter server (and reject any pending calls). */
  async stop(): Promise<void> {
    if (!this.started) return;
    // Detach the SSE listener so a long-lived `Patter` instance shared with
    // other consumers doesn't accumulate dead listeners every time a tool is
    // stopped/restarted.
    if (this.metricsStoreRef && this.sseListener) {
      this.metricsStoreRef.off('sse', this.sseListener);
    }
    this.sseListener = null;
    this.metricsStoreRef = null;
    // Drop any in-flight dial waiter (silently — caller will get the
    // shutdown rejection from `pending` below if it ever set its waiter).
    this.pendingDial = null;
    for (const [, p] of this.pending) {
      clearTimeout(p.timer);
      p.reject(new Error('PatterTool: shutdown while call pending'));
    }
    this.pending.clear();
    // Best-effort — Patter's `stop()` is on the embedded server; not all
    // versions expose a public stop on the Patter class.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const stoppable = this.phone as unknown as { stop?: () => Promise<void> };
    if (typeof stoppable.stop === 'function') {
      await stoppable.stop();
    }
    this.started = false;
  }

  // --- Execution ----------------------------------------------------------

  /** Place an outbound call and resolve once it ends with the transcript and metrics. */
  async execute(args: PatterToolExecuteArgs): Promise<PatterToolResult> {
    if (!this.started) await this.start();
    if (!args || typeof args.to !== 'string' || !args.to.startsWith('+')) {
      throw new Error('PatterTool.execute: `to` must be an E.164 phone number (e.g. "+15551234567").');
    }
    const timeoutSec = Math.max(
      5,
      Math.min(1800, args.max_duration_sec ?? this.maxDurationSec),
    );

    const baseAgent = this.agent ?? ({} as AgentOptions);
    const overrideAgent = this.phone.agent({
      ...baseAgent,
      ...(args.goal !== undefined ? { systemPrompt: args.goal } : {}),
      ...(args.first_message !== undefined ? { firstMessage: args.first_message } : {}),
    });

    // Serialize the dial → call_id capture across concurrent execute() calls.
    // `pendingDial` is a single slot, so two parallel callers would clobber
    // each other's resolver and one of them would hang forever waiting for
    // an SSE event that already went to the other. Chain through dialQueue
    // so exactly one execute() owns the slot at a time. Once the call_id is
    // captured we release the queue immediately — the actual call_end can
    // run concurrently with later dials.
    const callId = await this.acquireCallId(args.to, overrideAgent);

    return new Promise<PatterToolResult>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(callId);
        reject(new Error(`PatterTool.execute: call ${callId} exceeded ${timeoutSec}s timeout`));
      }, timeoutSec * 1000);
      this.pending.set(callId, {
        resolve,
        reject,
        timer,
        startedAt: Date.now() / 1000,
      });
    });
  }

  /** Issue the outbound dial under the mutex and return its assigned call_id. */
  private async acquireCallId(to: string, agent: AgentOptions): Promise<string> {
    // Chain on dialQueue. Each call replaces dialQueue with a tail promise so
    // the *next* execute() can also enqueue. We resolve the queue promise as
    // soon as we capture the call_id (or fail) so we don't hold the slot for
    // the call's full duration.
    let release!: () => void;
    const slot = new Promise<void>((r) => {
      release = r;
    });
    const previous = this.dialQueue;
    this.dialQueue = previous.then(() => slot);
    await previous;

    // We now own the slot. Set up the dispatcher, dial, and wait for the
    // call_initiated SSE — bounded by DIAL_CAPTURE_TIMEOUT_MS so a missed
    // event doesn't hang the caller forever.
    let captureTimer: NodeJS.Timeout | null = null;
    try {
      const callIdPromise = new Promise<string>((resolve, reject) => {
        this.pendingDial = resolve;
        captureTimer = setTimeout(() => {
          this.pendingDial = null;
          reject(
            new Error(
              `PatterTool.execute: did not observe call_initiated within ${PatterTool.DIAL_CAPTURE_TIMEOUT_MS}ms`,
            ),
          );
        }, PatterTool.DIAL_CAPTURE_TIMEOUT_MS);
      });

      await this.phone.call({
        to,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        agent: agent as any,
      });

      const callId = await callIdPromise;
      if (captureTimer) clearTimeout(captureTimer);
      return callId;
    } finally {
      // Always clear pendingDial; release the mutex so the next dial can run.
      if (captureTimer) clearTimeout(captureTimer);
      this.pendingDial = null;
      release();
    }
  }

  /**
   * Hermes-style handler: `(args, kwargs) => Promise<string>` returning a JSON
   * string with either the result envelope or an `{"error": "..."}` payload.
   * Mirrors the Python `PatterTool.hermes_handler` so cross-SDK adapters share
   * the same wire contract.
   */
  hermesHandler(): (args: PatterToolExecuteArgs) => Promise<string> {
    return async (args: PatterToolExecuteArgs) => {
      try {
        const result = await this.execute(args);
        return JSON.stringify(result);
      } catch (err) {
        return JSON.stringify({ error: err instanceof Error ? err.message : String(err) });
      }
    };
  }

  // --- Internal: onCallEnd dispatcher -------------------------------------

  private async onCallEndHandler(data: Record<string, unknown>): Promise<void> {
    const callId = (data.call_id as string) || '';
    if (!callId) return;
    const pending = this.pending.get(callId);
    if (!pending) {
      this.bus.emit('orphan_end', { call_id: callId, data });
      return;
    }
    clearTimeout(pending.timer);
    this.pending.delete(callId);
    const metrics =
      data.metrics && typeof data.metrics === 'object'
        ? (data.metrics as Record<string, unknown>)
        : null;
    const cost =
      metrics &&
      typeof metrics.cost === 'object' &&
      metrics.cost &&
      typeof (metrics.cost as Record<string, unknown>).total === 'number'
        ? ((metrics.cost as Record<string, unknown>).total as number)
        : undefined;
    const duration =
      typeof (metrics?.duration_seconds as number | undefined) === 'number'
        ? (metrics?.duration_seconds as number)
        : Math.max(0, Date.now() / 1000 - pending.startedAt);
    const transcript = Array.isArray(data.transcript)
      ? (data.transcript as PatterToolResult['transcript'])
      : [];
    const status = (data.status as string) || 'completed';
    pending.resolve({
      call_id: callId,
      status,
      duration_seconds: duration,
      cost_usd: cost,
      transcript,
      metrics,
    });
  }
}

File: libraries/python/getpatter/integrations/patter_tool.py
"""PatterTool — wrap a live Patter instance as a tool callable from external
agent frameworks (OpenAI Assistants, Anthropic Claude tool-use, LangChain,
Hermes Agent, MCP, generic OpenAI-compatible endpoints).

See ``libraries/typescript/src/integrations/patter-tool.ts`` for the matching TS module —
the wire contracts (parameter schema, result envelope, ``hermes_handler``
return shape) are kept identical so a customer can swap SDKs at any time.

Usage (Hermes Agent integration)::

    # in your tools/patter.py file (auto-discovered by Hermes)
    from tools.registry import registry
    from getpatter import Patter, Twilio, DeepgramSTT, GroqLLM, ElevenLabsTTS
    from getpatter.integrations import PatterTool

    phone = Patter(
        carrier=Twilio(),
        phone_number=os.environ["TWILIO_PHONE_NUMBER"],
        webhook_url="agent.example.com",
    )

    tool = PatterTool(
        phone=phone,
        agent={"stt": DeepgramSTT(), "llm": GroqLLM(), "tts": ElevenLabsTTS()},
    )

    # One-line registration with Hermes' registry:
    tool.register_hermes(registry)

Usage (OpenAI / Anthropic / generic)::

    tool = PatterTool(phone=phone, agent={...})
    schemas = {"openai": tool.openai_schema(), "anthropic": tool.anthropic_schema()}

    # When the LLM emits a tool_call:
    result = await tool.execute(to="+15551234567", goal="Book a dentist appointment")
    # → {"call_id": "...", "status": "completed", "duration_seconds": 12.3, ...}
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import asdict, dataclass, field
from typing import Any, Awaitable, Callable

logger = logging.getLogger("getpatter.integrations.patter_tool")


# JSON-Schema for the call args. Identical wire shape across openai/anthropic/hermes.
_PARAMETERS_SCHEMA: dict[str, Any] = {
    "type": "object",
    "properties": {
        "to": {
            "type": "string",
            "description": (
                'Destination phone number in E.164 format (e.g. "+15551234567"). Required.'
            ),
        },
        "goal": {
            "type": "string",
            "description": (
                "What the agent should accomplish on the call. Becomes the in-call "
                "agent's system prompt for this single call."
            ),
        },
        "first_message": {
            "type": "string",
            "description": (
                "Optional first message the agent speaks when the callee answers. "
                "Defaults to a generic greeting."
            ),
        },
        "max_duration_sec": {
            "type": "integer",
            "description": (
                "Hard timeout for the call in seconds. Default 180. The call is "
                "force-ended at this deadline whether or not it has resolved."
            ),
            "minimum": 5,
            "maximum": 1800,
        },
    },
    "required": ["to"],
}

_DEFAULT_NAME = "make_phone_call"
_DEFAULT_DESCRIPTION = (
    "Place a real outbound phone call. Returns a JSON object with the full "
    "transcript, call status, duration in seconds, and cost. Use this when the "
    "user asks you to call someone, schedule appointments by phone, or otherwise "
    "reach a human via voice."
)


@dataclass
class PatterToolResult:
    """Structured result returned by ``PatterTool.execute``."""

    call_id: str
    status: str
    duration_seconds: float
    transcript: list[dict[str, Any]] = field(default_factory=list)
    cost_usd: float | None = None
    metrics: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class PatterTool:
    """Wrap a live ``Patter`` instance as a tool callable from external agent
    frameworks. Schema-stable across OpenAI / Anthropic / Hermes; ``execute()``
    runs an outbound dial and waits for completion.
    """

    def __init__(
        self,
        phone: Any,
        agent: dict[str, Any] | None = None,
        name: str = _DEFAULT_NAME,
        description: str = _DEFAULT_DESCRIPTION,
        max_duration_sec: int = 180,
        recording: bool = False,
    ) -> None:
        if phone is None:
            raise ValueError("PatterTool: `phone` (a Patter instance) is required.")
        self._phone = phone
        self._agent_spec = agent
        self.name = name
        self.description = description
        self._max_duration_sec = max(5, min(1800, int(max_duration_sec)))
        self._recording = recording
        self._started = False
        # Map of call_id -> asyncio.Future awaiting the call_end payload.
        self._pending: dict[str, asyncio.Future] = {}
        # Single-slot future for the next dial's call_id (FIFO via Lock).
        self._next_call_id: asyncio.Future[str] | None = None
        self._dial_lock: asyncio.Lock | None = None

    # --- Schema exporters --------------------------------------------------

    def openai_schema(self) -> dict[str, Any]:
        """OpenAI Chat Completions / Assistants tool spec."""
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": _PARAMETERS_SCHEMA,
            },
        }

    def anthropic_schema(self) -> dict[str, Any]:
        """Anthropic Messages API tool spec."""
        return {
            "name": self.name,
            "description": self.description,
            "input_schema": _PARAMETERS_SCHEMA,
        }

    def hermes_schema(self) -> dict[str, Any]:
        """Hermes Agent (Nous Research) registry schema. Same JSON-Schema as
        Anthropic's, exposed under ``parameters`` to match Hermes' contract."""
        return {
            "name": self.name,
            "description": self.description,
            "parameters": _PARAMETERS_SCHEMA,
        }

    # --- Hermes registry helper -------------------------------------------

    def register_hermes(self, registry: Any, toolset: str = "patter") -> None:
        """Register this tool with a Hermes ``tools.registry.Registry`` instance.

        Hermes' tool contract is ``handler(args: dict, **kw) -> str`` returning
        a JSON string (errors as ``{"error": "..."}``). We bridge it to
        ``execute()``.
        """
        if not hasattr(registry, "register"):
            raise TypeError(
                "PatterTool.register_hermes: argument must be a Hermes Registry "
                "(needs a `.register()` method)."
            )
        handler = self.hermes_handler()
        registry.register(
            name=self.name,
            toolset=toolset,
            schema=self.hermes_schema(),
            handler=handler,
        )

    # --- Lifecycle ---------------------------------------------------------

    async def start(self) -> None:
        """Start the underlying Patter server. Idempotent."""
        if self._started:
            return
        if not self._agent_spec:
            raise ValueError(
                "PatterTool.start: `agent` config is required. Pass "
                "`{'stt': ..., 'llm': ..., 'tts': ...}` or an `engine` "
                "(e.g. OpenAIRealtime) when constructing PatterTool."
            )
        self._dial_lock = asyncio.Lock()
        built_agent = self._phone.agent(**self._agent_spec)
        await self._phone.serve(
            agent=built_agent,
            recording=self._recording,
            on_call_end=self._on_call_end,
        )
        # Subscribe to the metrics store SSE stream so we can correlate
        # outbound dials (`call_initiated`) with the call_id Patter assigns
        # at dial time.
        store = self._phone.metrics_store
        if store is None:
            raise RuntimeError(
                "PatterTool.start: phone.metrics_store is None after serve() — "
                "is the dashboard disabled?"
            )
        # The Python MetricsStore exposes an asyncio.Queue subscription.
        queue = store.subscribe()
        asyncio.create_task(self._consume_metrics_events(queue, store))
        self._started = True

    async def stop(self) -> None:
        """Best-effort shutdown — fail any pending calls and stop the server."""
        if not self._started:
            return
        for fut in list(self._pending.values()):
            if not fut.done():
                fut.set_exception(RuntimeError("PatterTool: shutdown while call pending"))
        self._pending.clear()
        if hasattr(self._phone, "stop"):
            try:
                await self._phone.stop()
            except Exception:  # pragma: no cover - defensive
                logger.debug("PatterTool.stop: phone.stop() failed", exc_info=True)
        self._started = False

    # --- Execution ---------------------------------------------------------

    async def execute(
        self,
        to: str,
        goal: str | None = None,
        first_message: str | None = None,
        max_duration_sec: int | None = None,
    ) -> PatterToolResult:
        """Dial outbound, wait for the call to end, return a structured result."""
        if not isinstance(to, str) or not to.startswith("+"):
            raise ValueError(
                'PatterTool.execute: `to` must be an E.164 phone number (e.g. "+15551234567").'
            )
        if not self._started:
            await self.start()
        timeout = max(5, min(1800, int(max_duration_sec or self._max_duration_sec)))

        agent_kwargs = dict(self._agent_spec or {})
        if goal is not None:
            agent_kwargs["system_prompt"] = goal
        if first_message is not None:
            agent_kwargs["first_message"] = first_message
        override_agent = self._phone.agent(**agent_kwargs)

        # Acquire the dial lock so concurrent execute() calls don't fight over
        # which one captures the next call_initiated event. Use try/finally
        # to guarantee `_next_call_id` is cleared on every exit path —
        # otherwise a TimeoutError leaves a completed-with-exception future
        # in place, and the SSE consumer task would call `set_result` on it
        # (raising InvalidStateError, silently swallowed) while the next
        # legitimate execute() would wait on a stale future.
        assert self._dial_lock is not None
        async with self._dial_lock:
            loop = asyncio.get_running_loop()
            self._next_call_id = loop.create_future()
            try:
                await self._phone.call(to=to, agent=override_agent)
                call_id: str = await asyncio.wait_for(
                    self._next_call_id, timeout=10.0
                )
            finally:
                self._next_call_id = None
            end_future: asyncio.Future = loop.create_future()
            self._pending[call_id] = end_future

        try:
            data = await asyncio.wait_for(end_future, timeout=timeout)
        except asyncio.TimeoutError as exc:
            self._pending.pop(call_id, None)
            raise TimeoutError(
                f"PatterTool.execute: call {call_id} exceeded {timeout}s timeout"
            ) from exc

        return _build_result(call_id, data)

    def hermes_handler(self) -> Callable[..., Awaitable[str]]:
        """Return a Hermes-compatible handler ``(args, **kw) -> Awaitable[str]``.

        The handler returns a JSON string with the result envelope, or
        ``{"error": "..."}`` if execution failed. Matches Hermes' contract
        (see ``hermes-agent.nousresearch.com/docs/developer-guide/adding-tools``).
        """

        async def handler(args: dict[str, Any], **_kw: Any) -> str:
            try:
                result = await self.execute(
                    to=args.get("to") or "",
                    goal=args.get("goal"),
                    first_message=args.get("first_message"),
                    max_duration_sec=args.get("max_duration_sec"),
                )
                return json.dumps(result.to_dict(), default=str)
            except Exception as exc:
                return json.dumps({"error": str(exc)})

        return handler

    # --- Internal: SSE consumer + onCallEnd hook --------------------------

    async def _consume_metrics_events(self, queue: asyncio.Queue, store: Any) -> None:
        try:
            while True:
                event = await queue.get()
                if (
                    event.get("type") == "call_initiated"
                    and self._next_call_id is not None
                    and not self._next_call_id.done()
                ):
                    call_id = event.get("data", {}).get("call_id") or ""
                    if call_id:
                        self._next_call_id.set_result(call_id)
        except asyncio.CancelledError:
            pass
        finally:
            try:
                store.unsubscribe(queue)
            except Exception:
                pass

    async def _on_call_end(self, data: dict[str, Any]) -> None:
        call_id = data.get("call_id") or ""
        if not call_id:
            return
        future = self._pending.pop(call_id, None)
        if future is None or future.done():
            return
        future.set_result(data)


def _build_result(call_id: str, data: dict[str, Any]) -> PatterToolResult:
    """Translate Patter's ``onCallEnd`` payload into the public envelope."""
    metrics = data.get("metrics") if isinstance(data.get("metrics"), dict) else None
    cost: float | None = None
    duration: float = 0.0
    if metrics is not None:
        # `metrics` may be a dataclass-like object or a plain dict depending on
        # how the underlying transport serialised it.
        cost_obj = metrics.get("cost") if isinstance(metrics, dict) else None
        if isinstance(cost_obj, dict) and isinstance(cost_obj.get("total"), (int, float)):
            cost = float(cost_obj["total"])
        dur_raw = metrics.get("duration_seconds") if isinstance(metrics, dict) else None
        if isinstance(dur_raw, (int, float)):
            duration = float(dur_raw)
    transcript = data.get("transcript") if isinstance(data.get("transcript"), list) else []
    return PatterToolResult(
        call_id=call_id,
        status=str(data.get("status") or "completed"),
        duration_seconds=duration,
        cost_usd=cost,
        transcript=list(transcript),
        metrics=metrics,
    )
